make file submissions dis/allowable

wbaid opened the pull request:
## Status

Ready for review
* See next section for tests outstanding

### Status of requested changes (since b5d493c)

1. [x] fix: "Error strings [...], and also in the inline comment
   `securedrop/source_templates/lookup.html`, which could be confusing
   to sources." ([1](https://github.com/freedomofpress/securedrop/pull/4879#pullrequestreview-305238244), [2](https://github.com/freedomofpress/securedrop/pull/4879#issuecomment-545709292))

2. [x] fix: "Bug: textarea width on Source Interface"

3. [x] refactor: "Proposed alternative language" for "Instance Configuration" page
    1. [x] fix: "UX issue: 'Update' button on Instance Configuration page"
    2. [x] fix: "UX issue: 'Document Uploads', 'Allow' language on Instance
       Configuration page"

4. [x] refactor: versioned `instance_config`
    1. [x] "one configuration option per column"
    2. [x] migration to add column sets default value
    3. [x] "use the configuration option that has null `valid_until`. When we
       update a config, we store the historical configuration entry by
       setting `valid_until=datetime.datetime.utcnow()`, then store the new
       configuration with null `valid_until`."
    - [x] test: migrations

5. [x] test: integration test


## Description of Changes

1. [x] `InstanceConfig` versioned key-value store [as outlined](https://github.com/freedomofpress/securedrop/pull/4879#discussion_r339185352)
    - [x] `load_instance_config()` sets each app's `app.instance_config`
    (via `@before_request`)

2. [x] Source interface checks
   `app.instance_config.allow_document_uploads`; if `False`:
    1. `/lookup` hides the file input and
    2. changes its heading to "Submit Messages" (rather than "Submit
       Files or Messages") and other strings accordingly; and
    3. `/submit` skips processing of `request.files`.
    4. `/metadata` returns this setting as `allow_document_uploads`.

3. [x] Journalist interface:
    1. `/admin/config` adds a section "Submission Preferences"; and
    2. `/admin/update-submission-preferences` updates
       `InstanceConfig.allow_document_uploads`.


## Testing

Test by toggling the **Prevent sources from uploading documents**
checkbox in the "Submission Preferences" section of the "Instance
Config" view.

1. If unchecked (default):  Observe no changes from current behavior.

2. If checked:  Observe the changes described in (2) and (3) above.


## Deployment

No action required.  After upgrading, administrators may use the
"Submission Preferences" section of the "Instance Config" view.


## Checklist

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the
  development container
    - [x] Fix `tests/test_i18n.py::test_verify_default_locale_en_us_if_not_defined_in_config` (expected by end of day Monday, October 21)

zenmonkeykstop commented:
Hi @wbaid - it looks like having this configured via the web interface is the preferred option for the core team. We'll be entering feature freeze for `v1.1.0` tomorrow so will be unlikely to be able to suggest or incorporate any changes for that version, but this should be on the list of changes for the following `v1.2.0` release.

If you have availability and would like to talk about the change, the Securedrop gitter forum is active at https://gitter.im/freedomofpress/securedrop , and we host weekday public standup meetings at 10AM Pacific/1PM Eastern at https://meet.google.com/ekb-kkhf-mrk

wbaid commented:
Thanks, @zenmonkeykstop.  We'll be happy to revise this pull request for
v1.2.

We took the approach of modifying "config.py" via the Ansible workflow
because there did not seem to be an obvious precedent in the journalist
interface's "admin.py" for configuration changes, as opposed to writes
to definite filesystem paths (the logo) or database updates (users).
Would you like the Web-based setting to take the form of:

1. an update to a simple key-value store defined as a SQLAlchemy model;

2. a write to (e.g.) an optional "config.json" or "config.yml"; or

3. something else we may be overlooking?

redshiftzero commented:
I agree with @zenmonkeykstop that this seems like a useful feature. Doing 1 (adding a new table called e.g. `instance_config` that we can use to store key/value config values) I think is the best way to go here. In that scenario the update process from the admin's perspective is better since:

1. immediate, i.e. no Ansible playbook run required, 
2. has nicer UX since using the CLI is not needed, 
3. and it's easier to coordinate changes in the case of multiple admins.

wbaid commented:
Thanks, @redshiftzero.  We'll revise accordingly.  We've marked this
pull request back to "work in progress" and will comment here once it's
been updated for the team to revisit.

eloquence commented:
Thanks @wbaid -- we've set aside time in our current sprint (10/9 to 10/23) to help get this over the finish line if you have time to work on it from your end.

wbaid commented:
@eloquence, we'll aim for the middle of next week.  Appreciate your
making time for it in the sprint.

wbaid commented:
With our apologies for the delay, this should be ready for review by
Monday.

wbaid commented:
Our updated branch is pushed and ready for your review, with the changes
re-summarized above.

We have not added a functional test for the source's /lookup view after
an admin has set ALLOW_DOCUMENT_UPLOADS = False.  We can add this this
week if you'd like.

emkll commented on securedrop/tests/migrations/migration_fef03c1ec779.py:
```diff
@@ -0,0 +1,40 @@
+
+            # The SQLite driver appears to return this rather than the
+            # expected NoSuchTableError.
+            except OperationalError:
```
I also see this, NoSuchTableError is treated as an OperationalError:
```
    def do_execute(self, cursor, statement, parameters, context=None):
>       cursor.execute(statement, parameters)
E       sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: instance_config
E       [SQL: SELECT * FROM instance_config]
E       (Background on this error at: http://sqlalche.me/e/e3q8)

/opt/venvs/securedrop-app-code/lib/python3.5/site-packages/sqlalchemy/engine/default.py:552: OperationalError
```

Since the downgrade is a drop table action, I think this is fine.

eloquence commented:
![Screen Shot 2019-10-22 at 16 56 58-fullpage](https://user-images.githubusercontent.com/213636/67344590-4e530180-f4ed-11e9-870e-1e8424900b22.png)

Screenshot of the admin interface, w/ new section highlighted in red, for UX feedback by @ninavizz.

eloquence commented:
A bit more UX feedback after discussion w/ @ninavizz. Thanks again for working on this.

## Bug: textarea width on Source Interface

The textarea currently remains at a fixed width if uploads are enabled. It should span the whole width of its container when this is the case.

![Screen Shot 2019-10-22 at 17 47 35-fullpage](https://user-images.githubusercontent.com/213636/67346640-3df25500-f4f4-11e9-94ee-a8aa1d5006e0.png)
![Screen Shot 2019-10-22 at 17 48 18-fullpage](https://user-images.githubusercontent.com/213636/67346639-3df25500-f4f4-11e9-9bd4-c357ddaff13c.png)

## UX issue: "Update" button on Instance Configuration page

Users may click the "Update" button expecting that it will change _all_ preferences, but it only changes a single preference. It is better to use more specific language on the button to avoid confusion.

## UX issue: "Document Uploads", "Allow" language on Instance Configuration page

This language is a bit awkward, and starting the option with "Allow" is potentially alarming ("am I allowing something I shouldn't be")? Virtually all SecureDrop users will want to use the default setting, so we want to make it easier for them to quickly recognize that this option is not one they need to worry about.

# Proposed alternative language
We propose the following 

```
Submission Preferences

[ ] Prevent sources from uploading documents
    Sources will still be able to send messages.

[[ Update Submission Preferences ]]
```

The text "Sources will still be able to send messages" should ideally be aligned as shown here, to be clearly understood as explanatory. However, if you run into difficulties here, you can also just make it two sentences ("Prevent sources from uploading documents. Sources will still be able to send messages.")

Hope that makes sense, happy to discuss!

wbaid commented:
@emkll and @eloquence (and @ninavizz):  Thank you for your feedback.  We
won't be able to complete all of these changes this week, but we'll
offer a revised branch for your review by the start of the week of
November 4.

wbaid commented on securedrop/models.py:
```diff
@@ -739,3 +739,15 @@ class RevokedToken(db.Model):
+
+    __tablename__ = 'instance_config'
+    name = Column(String, primary_key=True)
+    value = Column(JSON)
```
We chose `JSON` less to be able to represent a hierarchy than to entrust to SQLAlchemy the de/serialization of non-string values (e.g., the Boolean `allow_document_uploads`) without either:

1. inviting any risky use of (e.g.) `eval()`; or
2. constraining the typing of any future `instance_config` rows.

Would you prefer another approach?

redshiftzero replied:
@wbaid you're making a good point re: the typing of future configuration values, so we had a bit of a debate about the options here. For posterity here are the options we considered:

1. The table contains name/value pairs and one configuration option per row in the database table. This is a straightforward implementation but as you point out we can’t store the values with their proper types.
2. The table has one row for the entire configuration, and there is one configuration option per column. Also straightforward, we can store configuration with their correct type. The disadvantage here is that adding any new configuration values mean adding a new column (i.e. a db migration).
3. We store entire config as JSON as implemented in the PR. There's one row for the config. Configuration is stored as the proper type, but it's a bit harder to work with, i think we’d need to install an extension (https://www.sqlite.org/json1.html). This is important in the future where we might want to e.g. migrate some of the config data in the database instance config using alembic.

After discussion, I think 2 seems like the best approach. @rmol pointed out that option 2 also easily enables us to store the _history_ of the configuration for debugging issues: his idea was that we use the configuration option that has null `valid_until`. When we update a config, we store the historical configuration entry by setting `valid_until=datetime.datetime.utcnow()`, then store the new configuration with null `valid_until`. What do you think?

Let us know if you want help implementing any of these changes, i.e. if you'd prefer a maintainer to add a commit implementing this on top of your other changes here so we can merge your contribution.

wbaid replied:
Thanks, @redshiftzero.  This seems like a good approach.  We'll be happy
to include this refactoring along with the other changes requested.

ninavizz commented:
One tiiny thing @eloquence did not include but we had discussed: in the line of text on the "Submission" page in the Source UI that reads:
`If you are already familiar with GPG, you can optionally encrypt your files and messages with our public key`

...the bit of text: `files and` should be removed. Speaking to files on this page is rather awkward and odd, when the goal is to restrict users to messages, only.

As such, it should read: `If you are already familiar with GPG, you can optionally encrypt your messages with our public key...`

I don't know if this should be punted into the Localization translator team's box, or if this is something y'all can handle in this PR. @rmol or @redshiftzero thoughts?

wbaid commented:
Thanks, @redshiftzero.  This seems like a good approach.  We'll be happy
to include this refactoring along with the other changes requested.

wbaid commented:
See new HEAD, with changes summarized (and revisions outlined) above.
Other notes:

1. Migration tests and an integration test will be forthcoming this
   week, pending any further feedback.

2. We have appended these revisions as new commits, on the assumption
   that once approved this pull request will be squash-merged.  Let us
   know if you'd like us to rebase things more logically for review
   purposes.

3. Even with the inversion of the admin UX to "*Prevent* sources...",
   we have kept the *setting* as `allow_document_uploads` so that
   conditionals elsewhere remain intuitive: e.g. "if allow..." rather
   than "if not prevent...".

4. app.instance_config is now completely distinct from
   app.{config,sdconfig}: a separate object/attribute rather than an
   extra set of keys to be inserted at runtime.

wbaid commented:
FYI:  We won't be available to work further on this pull request until
the week of November 18.  We look forward to your feedback.

wbaid commented:
@emkll or @eloquence, will the team still be able to review this pull
request in time for inclusion in v1.2.0?  We saw the announcement of
today's feature-level freeze and can work with you to finalize this as
quickly as possible this week, pending your feedback.

redshiftzero commented:
heads up @wbaid it looks like there are some alembic tests failures occurring in CI related to this diff

redshiftzero commented on securedrop/alembic/versions/e0719fbc39a0_add_instance_config_allow_document_.py:
```diff
@@ -0,0 +1,40 @@
+    # instance_config setting (column), all we have to do is insert a
+    # row with its default value.
+    conn = op.get_bind()
+    conn.execute("""INSERT INTO instance_config (allow_document_uploads) VALUES (1)""")
```
:+1:

wbaid commented:
Many thanks, @redshiftzero.  We've squashed the Alembic migrations and
fixed the failing migration tests.  Thank you for pointing out this
oversight on our part.

We wanted to wait for this high-level confirmation of the new
instance_config functionality before we add final tests for the
allow_document_uploads setting.  Would you like to see an integration
test (per @emkll[1]), a functional test, or both?  We can have either or
both tests pushed tonight in order to meet the EOD deadline.


[1]:
https://github.com/freedomofpress/securedrop/pull/4879#pullrequestreview-305238244

redshiftzero commented:
oh nice thanks for the fast update! Just the integration and unit coverage is required for merge given all the work you've put in here. Let's consider the functional/selenium based tests as nice to have and we can add as a followup issue for another contributor

wbaid commented:
Thanks, @redshiftzero!  We've just pushed integration tests for
allow_document_uploads={True,False} to accompany the unit tests in
"test_journalist.py" and "test_source.py".

We appreciate your looking at this today.

eloquence commented:
Thanks for making all the UX changes previously discussed, @wbaid -- as @redshiftzero noted, it looks great from a UX perspective! The only thing we may want to track as a follow-up enhancement is adding a flashed message after you click "Update Submission Preferences", to let the user know that the settings have been successfully modified. But we can tackle this as a follow-up issue.

wbaid commented:
Thanks to all---@redshiftzero, @eloquence, @emkll, @ninavizz,
@zenmonkeykstop---for your feedback and your time in getting this ready
for merge.
